File: gsm-symbolic-reproduction/symbolize_gsm8k.py
import json
import sys
import subprocess

def run_command(command):
    process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()
    if process.returncode != 0:
        print(f"Error executing command: {command}", file=sys.stderr)
        print(f"Error message: {stderr.decode('utf-8')}", file=sys.stderr)
        sys.exit(1)
    return stdout.decode('utf-8')

def convert_problem(question, answer):
    prompt = json.dumps({"question": question, "answer": answer})
    
    template_command = f"template-populator -template convert.prompt.txt -strict -verbose '{prompt}'"
    template_output = run_command(f"{template_command}")
    print(f"Template output: {template_output}", file=sys.stderr)

    import time
    time.sleep(4)
    
    # Escape special characters and newlines for shell
    escaped_template_output = template_output.replace("'", "'\\''").replace('\n', '\\n')
    
    cgpt_output = run_command(f"cgpt -i '{escaped_template_output}'")
    print(f"CGPT output: {cgpt_output}", file=sys.stderr)
    
    # Escape special characters for echo
    escaped_cgpt_output = cgpt_output.replace("'", "'\\''")
    
    xq_command = "xq -j"
    xq_output = run_command(f"echo '{escaped_cgpt_output}' | {xq_command}")
    print(f"XQ output: {xq_output}", file=sys.stderr)
    
    # Escape special characters for echo
    escaped_xq_output = xq_output.replace("'", "'\\''")
    
    jq_command = "jq -c '{question: .symbolic_template_question, answer: .symbolic_template_answer}'"
    final_output = run_command(f"echo '{escaped_xq_output}' | {jq_command}")
    print(f"Final output: {final_output}", file=sys.stderr)
    
    return final_output

def main():
    # print to stderr:
    print("Starting conversion...", file=sys.stderr)
    for line in sys.stdin:
        print(line, file=sys.stderr)
        data = json.loads(line)
        result = convert_problem(data['question'], data['answer'])
        sys.stdout.write(result)

if __name__ == "__main__":
    main()

File: gsm-symbolic-reproduction/symbolize_gsm8k.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
)

type QAPair struct {
	Question string `json:"question"`
	Answer   string `json:"answer"`
}

func runCommand(command string, args ...string) (string, error) {
	cmd := exec.Command(command, args...)
	cmd.Stderr = os.Stderr
	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("error executing command '%s': %w", command, err)
	}
	return string(output), nil
}

func convertProblem(rowNumber int, question, answer string) (string, error) {
	example := QAPair{
		Question: question,
		Answer:   answer,
	}
	prompt, err := json.Marshal(example)
	if err != nil {
		errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": example})
		fmt.Printf("\n%s\n", errorInput)
		return "", fmt.Errorf("error marshalling example: %w", err)
	}

	templateOutput, err := runCommand("template-populator", "-template", "convert.prompt.txt", "-strict", "-verbose", string(prompt))
	if err != nil {
		errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": string(prompt)})
		fmt.Printf("\n%s\n", errorInput)
		return "", fmt.Errorf("template command failed: %w", err)
	}

	cgptCmd := exec.Command("cgpt")
	cgptCmd.Stdin = strings.NewReader(templateOutput)
	cgptCmd.Stderr = os.Stderr
	cgptOutput, err := cgptCmd.Output()
	if err != nil {
		errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": templateOutput})
		fmt.Printf("\n%s\n", errorInput)
		return "", fmt.Errorf("cgpt command failed: %w", err)
	}

	xqCmd := exec.Command("xq", "-j")
	xqCmd.Stdin = strings.NewReader(string(cgptOutput))
	xqCmd.Stderr = os.Stderr
	xqOutput, err := xqCmd.Output()
	if err != nil {
		errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": string(cgptOutput)})
		fmt.Printf("\n%s\n", errorInput)
		return "", fmt.Errorf("xq command failed: %w", err)
	}

	jqCmd := exec.Command("jq", "-c", "{question: .symbolic_template_question, answer: .symbolic_template_answer}")
	jqCmd.Stdin = bytes.NewReader(xqOutput)
	jqCmd.Stderr = os.Stderr
	finalOutput, err := jqCmd.Output()
	if err != nil {
		errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": string(xqOutput)})
		fmt.Printf("\n%s\n", errorInput)
		return "", fmt.Errorf("jq command failed: %w", err)
	}

	output := string(finalOutput)
	return strings.TrimSpace(output), nil
}

func main() {
	logger := log.New(os.Stderr, "", log.LstdFlags)
	logger.Println("Starting conversion...")
	scanner := bufio.NewScanner(os.Stdin)
	rowNumber := 0
	for scanner.Scan() {
		rowNumber++
		line := scanner.Text()
		logger.Println(line)
		var data map[string]string
		if err := json.Unmarshal([]byte(line), &data); err != nil {
			logger.Printf("Error parsing JSON: %v\n", err)
			errorInput, _ := json.Marshal(map[string]interface{}{"row": rowNumber, "input": line})
			fmt.Printf("\n%s\n", errorInput)
			continue
		}
		result, err := convertProblem(rowNumber, data["question"], data["answer"])
		if err != nil {
			logger.Printf("Error converting problem: %v\n", err)
			continue
		}
		fmt.Println(result)
	}
	if err := scanner.Err(); err != nil {
		logger.Fatalf("Error reading input: %v\n", err)
	}
}
